crates/utils/re_uri/src/origin.rs

```diff
@@ -48,25 +48,49 @@ impl Origin {
 
     /// Parses a URL and returns the [`crate::Origin`] and the canonical URL (i.e. one that
     ///  starts with `http://` or `https://`).
-    pub(crate) fn replace_and_parse(value: &str) -> Result<(Self, url::Url), Error> {
-        let scheme: Scheme = value.parse()?;
-        let rewritten = scheme.canonical_url(value);
+    pub(crate) fn replace_and_parse(input: &str) -> Result<(Self, url::Url), Error> {
+        let scheme: Scheme;
+        let rewritten;
+
+        if !input.contains("://") && (input.contains("localhost") || input.contains("127.0.0.1")) {
+            // Assume `rerun+http://`, because that is the default for localhost
+            scheme = Scheme::RerunHttp;
+            rewritten = format!("http://{input}");
+        } else {
+            scheme = input.parse()?;
+            rewritten = scheme.canonical_url(input);
+        }
 
         // We have to first rewrite the endpoint, because `Url` does not allow
         // `.set_scheme()` for non-opaque origins, nor does it return a proper
         // `Origin` in that case.
         let mut http_url = url::Url::parse(&rewritten)?;
 
-        // If we parse a Url from e.g. `https://redap.rerun.io:443`, `port` in the Url struct will
-        // be `None`. So we need to use `port_or_known_default` to get the port back.
-        // See also: https://github.com/servo/rust-url/issues/957
-        if http_url.port_or_known_default().is_none() {
-            // If no port is specified, we assume the default redap port:
-            http_url.set_port(Some(51234)).ok();
+        let mut has_port = true;
+        {
+            // Parsing with a non-standard scheme
+            // is a hack to work around the `url` crate bug
+            // <https://github.com/servo/rust-url/issues/957>.
+            if let Some(rest) = http_url.to_string().strip_prefix("http://") {
+                has_port = url::Url::parse(&format!("foobarbaz://{rest}"))?
+                    .port()
+                    .is_some();
+            } else if let Some(rest) = http_url.to_string().strip_prefix("https://") {
+                has_port = url::Url::parse(&format!("foobarbaz://{rest}"))?
+                    .port()
+                    .is_some();
+            } else {
+                // Should not happen.
+            };
+        }
+
+        if !has_port {
+            // If no port is specified, we assume the default:
+            http_url.set_port(Some(crate::DEFAULT_PROXY_PORT)).ok();
```

Comment:
Wild! 🤓

Reply:
yes, that would be nice, but since `Origin` doesn't know about that it would require more structural changes (e.g. have `Origin` parse port to `None`).

It would be easier to just switch redap to use the same port as OSS

Reply:
It doesn't make sense for `Origin` to even be able to produce a default port. Only `RedapUri` has sufficient context for that.

Reply:
> It would be easier to just switch redap to use the same port as OSS

It can't be on the same port until it's the same service

Reply:
…or else?

Reply:
> as long as we don't run both at the same time on the same machine

You are doing that every time you run dataplatform locally at the same time as a native viewer, unless the viewer is started with `--connect`. Even if the proxy isn't used, it still starts up.